### DISCUSSION TITLE
Create PDF by streaming data

### DISCUSSION THREAD

**Discussion by loust333** [gettalong/hexapdf#310] [Q&A] [answered]
Hello @gettalong  😊

I was playing around with the plugin. I have a use case where a PDF for the entire year needs to be generated which contains every employee and data about their payslips. 

Is there a way to progressively write in a StringIO or TempFile and, after writing, resetting the Composer (keeping styles but forgetting pages) ? And afterwards to concatenate the new pages to the Tempfile ? 

I tried to merge, but the issue is that it still loads all the objects using that method.. 😅

Did I miss something ? 

![Snip20240627_19](https://github.com/gettalong/hexapdf/assets/12141604/fb9decb6-0da4-4dc9-861b-8ff594d7f322)

**Comment by gettalong** [comment]
Hello @loust333,

It is currently not possible to do that kind of streaming because creating the pages is separate from writing the PDF.  However, it might be possible to save some memory by finalizing pages earlier. I will have to look at this to see how this could be done (similar to how forgetting the `Canvas` instances in the raw text benchmark - https://github.com/gettalong/hexapdf/blob/master/benchmark/raw_text/hexapdf.rb#L18-L21 - works).

Do you have some generalized example code for your use-case that you can share?

**Reply by gettalong** [reply]
I did some test with the line wrapping benchmark script. When it does `composer.text(whole_file)` it consumes about 4 times the memory as when the `whole_file` is split into smaller chunks and each of them processed one after the other. So it might help if output each part as soon as it is available and not build up a tree of box objects first.

**Reply by loust333** [reply]
Sure: 

I had fun creating this gist: https://gist.github.com/loust333/91a273aa5b568a51a795388603667a9a

The gist is what I tested for one report but it blocked with the memory, see the comments in the gist ("hard") option

I added in comments what I tried out for the progressive writing in comments. 😄 

Let me know if it helps. 

Here attached the generated dummy pdf 
[project_report_summary.pdf](https://github.com/user-attachments/files/16036592/project_report_summary.pdf)

**Reply by loust333** [reply]
@gettalong,   I used text and tables which made a lot of objects, you can see in the gist. That would explain the memory spike.

**Reply by gettalong** [reply]
Thanks @loust333! I will have a look at it over the weekend and see what I can do.

**Reply by gettalong** [reply]
Okay, I had a quick look at it:

* First of all: The report looks very nice! :+1:
* I changed a bunch of lines to remove Rails syntax but that shouldn't influence the result in any way.
* I was able to reproduce your results.
* However, using `Benchmark.memory` produces quite some overhead, so I commented out the lines and ran the script using `/usr/bin/time` to get runtime and memory usage. On difficulty hard (=30) it took 13s and 78MiB without YJIT and 7s and 88MiB with YJIT. The generated report has 271 pages and that memory usage seems okay to me in this case.
* Just to have something to compare with, I set the difficulty to 100. That generated a report with 2401 pages and a size of abou 26MiB, took with YJIT about 82s and used about 436MiB memory.

I will have a more in-depth look later to see if it is possible to optimize this.

**Reply by loust333** [reply]
That's weird. Without any changes, to hexapdf, it uses only that much memory. 

Can you share the code so that I can try out in rails ? Is it because of the benchmark memory ? 😵

**Reply by gettalong** [reply, marked as answer]
I just added the needed require statements and changed `#present?` and `#second` to their plain Ruby counterparts. That made the script run-able without Rails, giving your results. Then I commented out the `Benchmark.measure` lines (56/57 plus their respective `end` lines).

I haven't looked at 'benchmark-memory' but in my experience all memory measurement tools have some overhead, especially when many objects are involved as in this case. The difference in memory usage on the medium difficulty level was ~200MiB without Benchmark.memory and ~1400MiB with it.

**Reply by loust333** [reply]
Thanks for the explanation, i will try out and give you more feedback. 🙂

**Reply by loust333** [reply]
Hi @gettalong, 

I tested out, and it was indeed the Benchmark. memory which was creating the overhead. After removing, the generation of PDF was completed without any issue. 😅

Thanks a lot. I just need to test 8 more PDF documents. 😩

**Reply by gettalong** [reply]
Let me know if you need any help!

For what's worth:

* You can save some memory by creating a style for `:small_right`, i.e. `style(:small_right, base: :small, text_align: :right)` which saves some memory (on difficulty 10 it saved about a quarter of Style objects and 6000 less Arrays as well as 11000 less Hashes)

* You can also reduce memory and increase performance by removing the `cell_style` from `container.table(cells, cell_style: { border: { width: 0 }, padding: [1, 2 ,1, 1] }) do |args|` and moving it as first part into args, i.e. `args[0..-1, 0..-1] = { border: { width: 0 }, padding: [1, 2 ,1, 1] }`.

* Two small hints: You don't need to specify `style: :base` since that is the default. And if part of the data to `#formatted_text` looks like `{text: "..."}` you don't need the hash and can just use "...". For example, line 107 `{text: "#{employee_data[:name]} (", style: :base},` could also be written as `"#{employee_data[:name]} (",`.

* I will update HexaPDF to allow step sizes for ranges, i.e. `(1..11).step(2)`, when defining table cell styles. This allows you to define the table cell styles using just 4 instructions instead of repeating the same one for all uneven rows/columns).

* I found another small optimization for HexaPDF itself that saves memory and is more performant by not relying on `HexaPDF::Document::Layout#method_missing`.

**Reply by loust333** [reply]
Thanks a lot @gettalong. It worked like a charm.

I will continue testing the implementation of HexaPDF in all the others PDFs. 🥱

One little remark that I encounter during the test phase: 

```
cells << [
    layout.text(income_line[:code] || '', style: :small),
    layout.text(income_line[:amount] || '', style: :small),
  ]
  
composer.table(cells)
```

OR 

`composer.text(redundant_translations[:title], style: :h2)`

If the variable is nil an error is raised. Sometimes the variable may be nil. To avoid the issue || '' is needed in most calls.


We can keep this discussion open for those that test the methods with Benchmark.memory 😂

**Reply by gettalong** [reply]
There might be a difference between leaving out the call to `#text` completely or just passing an empty string, depending on the assigned style properties padding and border. So if in your case just using `|| ''` is okay, please do that. 

Here is an example where it makes a difference:

~~~ruby
require 'hexapdf'

HexaPDF::Composer.create('tb.pdf') do |c|
  data = nil
  c.text('test')
  c.text(data || '', padding: 10)
  c.text('test')
  c.text('test')
  c.text(data, padding: 10) if data
  c.text('test')
end
~~~

![image](https://github.com/gettalong/hexapdf/assets/4104/76a971fc-f68d-4b48-9e40-1ea280fe5181)

**Reply by loust333** [reply]
Ah ok, it's noted 😊
